rlapy/comps/evd.py

```diff
@@ -0,0 +1,381 @@
+        # indices of r largest components of |λ|
+        U = U[:,I]
+        lambda_matrix = lambda_matrix[I] 
+        V = Q @ U
+        lambda_matrix = np.diag(lambda_matrix)
+        return V, lambda_matrix
+
+class EVD2(EVDecomposer):
+
+    TOL_CONTROL = 'unknown'  # depends on implementation of rangefinder
+
+    def __init__(self, rs: RowSketcher):
+        """
+        Parameters
+        ----------
+        rs : RowSketcher
+        """
+        self.rowsketcher = rs
+
+    def __call__(self, A, k, epsilon, s, rng):
+        """
+        Rely on a RowSketcher to obtain the matrix S for the sketching matrix
+        S. Once we have S, we construct Y = A @ S and return Cholesky decomposition of 
+        S^T @ Y. This function is agnostic to the implementation of the
+        RowSketcher: it might has different ways of row sketching. We make no
+        assumptions on the RowSketcher's termination criteria beyond those
+        listed below.
+
+        Return the eigen decomposition matrices (V, lambda_matrix),
+        based on a row sketched version of A.
+        Use a Gaussian sketching matrix and pass over A a total of
+        num_passes times.
+
+        Parameters
+        ----------
+        num_passes : int
+            Total number of passes over A. We require num_passes >= 2.
+            
+        A : Union[ndarray, spmatrix, LinearOperator]
+            Data matrix to approximate. A must be an n × n Hermitian PSD matrix. But we did not check PSD
+
+        k : int
+            Target rank for the approximation of A: 0 < k < min(A.shape).
+            This parameter includes any oversampling. For example, if you
+            want to be near the optimal (Eckhart-Young) error for a rank 20
+            approximation of A, then you might want to set k=25.
+            
+        epsilon = np.nan : NaN by default
+            Target accuracy for the oversampled approximation of A: 0 < epsilon < np.inf.
+            This parameter inherits from the QBDecomposer or RangeFinder class.
+            
+        s  : int
+            Auxiliary parameter for the QBDecomposer or RangeFinder.
+
+        rng : Union[None, int, SeedSequence, BitGenerator, Generator]
+            Determines the numpy Generator object that manages randomness
+            in this function call.
+
+        Returns
+        -------
+        V : ndarray
+            Has shape (A.shape[0], k). Columns are orthonormal.
+
+        lambda_matrix : ndarray
+            Has shape (k,k). lambda_matrix is a diagonal matrix storing the eigenvalues of 
+            the matrix A
+
+        """
+        assert k > 0
+        assert k <= min(A.shape)
+        if not np.isnan(epsilon):
+            assert epsilon >= 0
+            assert epsilon < np.inf
+        rng = np.random.default_rng(rng)
+        S = self.rowsketcher(A, k + s,rng)
+        n = A.shape[0]
+        Y = A @ S
+        epsilon_mach = epsilon # a temporary regularization parameter
```

Comment:
``epsilon_mach`` should be set based on machine epsilon ``np.finfo(float).eps`` and the norm of ``Y``. It's unrelated to the user-specified accuracy parameter.